File: recompiler-rs/scripts/dump_autoroute.py
#!/usr/bin/env python3
"""Differential-oracle shim: run the cfg-mutating autoroute passes in the SAME
order `tools/v2_regen.py` does (wrapper -> tail_call -> pha_rts ->
dispatch-helper-discovery -> exit_mx) over the live cfg corpus, then dump the
resulting per-bank entry set + dispatch_helpers map + each pass's fix count as
JSON so the Rust `dump-autoroute` bin can be diffed against it.

The process-global decode state (inline-skip map + reloc regions) is installed
exactly as v2_regen installs it before the autoroute block, so the passes decode
under identical inputs on both sides.

Usage:
  dump_autoroute.py --rom sf.sfc --cfg-dir recomp --out golden/autoroute.json
"""
import argparse
import glob
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
SNESRECOMP = os.path.dirname(os.path.dirname(HERE))
sys.path.insert(0, os.path.join(SNESRECOMP, "recompiler"))
sys.path.insert(0, os.path.join(SNESRECOMP, "recompiler", "v2"))

from snes65816 import load_rom, set_active_reloc_regions  # noqa: E402
from v2.cfg_loader import load_bank_cfg  # noqa: E402
from v2.decoder import (  # noqa: E402
    classify_dispatch_helper, decode_function, set_global_inline_skip,
)
from v2.wrapper_autoroute import detect_and_route as autoroute_wrappers  # noqa: E402
from v2.tail_call_autoroute import detect_and_route as autoroute_tail_calls  # noqa: E402
from v2.pha_rts_autoroute import detect_and_route as autoroute_pha_rts  # noqa: E402
from v2.exit_mx_autoroute import detect_and_route as autoroute_exit_mx  # noqa: E402

_BANK_RE = __import__("re").compile(r"bank([0-9A-Fa-f]{2})\.cfg$")


def entry_obj(e):
    # exit_mx / inline_skip / force_variants are non-default dynamic attrs the
    # cfg loader only sets when the func line carries them.
    exit_mx = getattr(e, "exit_mx", None)
    inline_skip = getattr(e, "inline_skip", None)
    force_variants = getattr(e, "force_variants", None)
    return {
        "name": e.name,
        "start": e.start & 0xFFFF,
        "end": (e.end & 0xFFFF) if e.end is not None else None,
        "entry_m": int(e.entry_m),
        "entry_x": int(e.entry_x),
        "exit_mx": [int(exit_mx[0]), int(exit_mx[1])] if exit_mx is not None else None,
        "force_variants": (
            [[int(p[0]), int(p[1])] for p in force_variants]
            if force_variants else None
        ),
        "inline_skip": (int(inline_skip) if inline_skip is not None else None),
        "tail_call_pc16": (e.tail_call_pc16 & 0xFFFF) if e.tail_call_pc16 is not None else None,
        "entry_s_offset": int(e.entry_s_offset),
    }


def indirect_obj(d):
    return [
        d["site_pc16"] & 0xFFFF,
        int(d["count"]),
        str(d["idx_reg"]),
        [b & 0xFFFF for b in d["table_bases"]],
    ]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--rom", required=True)
    ap.add_argument("--cfg-dir", required=True)
    ap.add_argument("--out", required=True)
    args = ap.parse_args()

    rom = load_rom(args.rom)

    # Load cfgs in sorted-glob (== bank-ascending) order, mirroring v2_regen.
    parsed = []  # (bank, path, cfg)
    name_map = {}
    cross_bank_names = {}
    for p in sorted(glob.glob(os.path.join(args.cfg_dir, "bank*.cfg"))):
        m = _BANK_RE.search(os.path.basename(p))
        if not m:
            continue
        bank = int(m.group(1), 16)
        try:
            cfg = load_bank_cfg(p)
        except Exception as e:
            print(f"  parse-fail {p}: {e}", file=sys.stderr)
            continue
        parsed.append((bank, p, cfg))
        for entry in cfg.entries:
            if entry.name:
                name_map[(bank << 16) | (entry.start & 0xFFFF)] = entry.name
        for nd in cfg.names:
            addr = nd.addr_24 & 0xFFFFFF
            name_map[addr] = nd.name
            cross_bank_names.setdefault((addr >> 16) & 0xFF, []).append(nd)

    # Install process-global decode state as v2_regen does pre-autoroute.
    callee_inline_skip = {}
    for bank, _p, cfg in parsed:
        for entry in cfg.entries:
            n = getattr(entry, "inline_skip", None)
            if n:
                callee_inline_skip[(bank << 16) | (entry.start & 0xFFFF)] = n
    set_global_inline_skip(callee_inline_skip)

    reloc_regions = []
    for _bank, _p, cfg in parsed:
        if getattr(cfg, "reloc_regions", None):
            reloc_regions.extend(cfg.reloc_regions)
    set_active_reloc_regions(reloc_regions or None)

    # 1. wrapper
    wrapper_fixes = autoroute_wrappers(parsed, name_map, cross_bank_names, rom)
    # 2. tail_call
    tail_call_fixes = autoroute_tail_calls(parsed, rom)
    # 3. pha_rts
    pha_rts_fixes = autoroute_pha_rts(parsed, rom)
    # 4. dispatch-helper discovery
    dispatch_helpers = {}
    jsl_targets = set()
    for bank, _p, cfg in parsed:
        for entry in cfg.entries:
            try:
                g = decode_function(rom, bank, entry.start,
                                    entry_m=entry.entry_m, entry_x=entry.entry_x,
                                    end=entry.end)
            except Exception:
                continue
            for di in g.insns.values():
                ins = di.insn
                if ins.mnem == "JSL":
                    jsl_targets.add(ins.operand & 0xFFFFFF)
                elif ins.mnem == "JMP" and ins.length == 4:
                    jsl_targets.add(ins.operand & 0xFFFFFF)
    for tgt in jsl_targets:
        kind = classify_dispatch_helper(rom, (tgt >> 16) & 0xFF, tgt & 0xFFFF)
        if kind:
            dispatch_helpers[tgt] = kind
    # 5. exit_mx
    exit_mx_fixes = autoroute_exit_mx(parsed, rom, dispatch_helpers=dispatch_helpers)

    banks = {}
    for bank, _p, cfg in parsed:
        entries = sorted(
            (entry_obj(e) for e in cfg.entries),
            key=lambda o: (o["start"], o["name"] or ""),
        )
        per_variant = sorted(
            [int(t[0]), int(t[1]) & 0xFFFF, int(t[2]), int(t[3]), int(t[4]), int(t[5])]
            for t in cfg.exit_mx_at_per_variant
        )
        ind = sorted((indirect_obj(d) for d in cfg.indirect_dispatch),
                     key=lambda o: (o[0], o[1]))
        banks[str(bank)] = {
            "entries": entries,
            "exit_mx_at_per_variant": per_variant,
            "indirect_dispatch": ind,
        }

    out = {
        "fix_counts": {
            "wrapper": len(wrapper_fixes),
            "tail_call": len(tail_call_fixes),
            "pha_rts": len(pha_rts_fixes),
            "exit_mx": len(exit_mx_fixes),
        },
        "dispatch_helpers": {str(k): v for k, v in dispatch_helpers.items()},
        "banks": banks,
    }
    with open(args.out, "w") as f:
        json.dump(out, f, sort_keys=True, separators=(",", ":"))
    print(f"wrapper={len(wrapper_fixes)} tail_call={len(tail_call_fixes)} "
          f"pha_rts={len(pha_rts_fixes)} exit_mx={len(exit_mx_fixes)} "
          f"helpers={len(dispatch_helpers)} banks={len(banks)} -> {args.out}")


if __name__ == "__main__":
    main()

File: recompiler-rs/src/bin/dump_autoroute.rs
//! `dump-autoroute` — Rust side of the autoroute differential oracle.
//!
//! Loads the ROM + cfg dir, installs the same decode env state v2_regen does
//! pre-autoroute (aggregate reloc regions + inline-skip map), runs the five
//! cfg-mutating passes in v2_regen order (wrapper -> tail_call -> pha_rts ->
//! dispatch-helper-discovery -> exit_mx), and dumps the resulting per-bank entry
//! set + dispatch_helpers map + per-pass fix counts in the identical JSON shape
//! as `scripts/dump_autoroute.py`, for `scripts/diff_autoroute.py`.

use std::collections::HashMap;
use std::path::PathBuf;

use serde_json::{json, Map, Value};

use snesrecomp_regen::autoroute::{
    discover_dispatch_helpers, exit_mx_detect_and_route, pha_rts_detect_and_route,
    tail_call_detect_and_route, wrapper_detect_and_route,
};
use snesrecomp_regen::cfg::{load_bank_cfg, BankCfg, BankEntry, IndirectDispatch, NameDecl};
use snesrecomp_regen::decoder::DecodeEnv;
use snesrecomp_regen::rom::{load_rom, RelocRegion};

fn arg_value(args: &[String], flag: &str) -> Option<String> {
    args.iter().position(|a| a == flag).and_then(|i| args.get(i + 1).cloned())
}

/// Extract the bank from a `bankXX.cfg` filename (two hex digits), like the
/// `_BANK_CFG_RE` v2_regen uses.
fn filename_bank(p: &std::path::Path) -> Option<u32> {
    let name = p.file_name()?.to_str()?;
    let rest = name.strip_prefix("bank")?;
    let hex = rest.strip_suffix(".cfg")?;
    if hex.len() == 2 {
        u32::from_str_radix(hex, 16).ok()
    } else {
        None
    }
}

fn entry_obj(e: &BankEntry) -> Value {
    json!({
        "name": e.name,
        "start": e.start & 0xFFFF,
        "end": e.end.map(|v| v & 0xFFFF),
        "entry_m": e.entry_m,
        "entry_x": e.entry_x,
        "exit_mx": e.exit_mx.map(|(m, x)| json!([m, x])),
        "force_variants": e
            .force_variants
            .as_ref()
            .map(|v| Value::Array(v.iter().map(|&(m, x)| json!([m, x])).collect())),
        "inline_skip": e.inline_skip,
        "tail_call_pc16": e.tail_call_pc16.map(|v| v & 0xFFFF),
        "entry_s_offset": e.entry_s_offset,
    })
}

fn indirect_obj(d: &IndirectDispatch) -> Value {
    json!([
        d.site_pc16 & 0xFFFF,
        d.count,
        d.idx_reg.to_string(),
        d.table_bases.iter().map(|b| b & 0xFFFF).collect::<Vec<u32>>(),
    ])
}

fn main() {
    // Suppress panic spam from decode_function (caught and mapped to skips).
    std::panic::set_hook(Box::new(|_| {}));

    let args: Vec<String> = std::env::args().collect();
    let rom_path = arg_value(&args, "--rom").expect("--rom required");
    let cfg_dir = arg_value(&args, "--cfg-dir").expect("--cfg-dir required");
    let out_path = arg_value(&args, "--out").expect("--out required");

    let rom = load_rom(&rom_path).expect("load rom");

    let mut cfg_paths: Vec<PathBuf> = std::fs::read_dir(&cfg_dir)
        .expect("read cfg dir")
        .filter_map(|e| e.ok().map(|e| e.path()))
        .filter(|p| {
            p.file_name()
                .and_then(|n| n.to_str())
                .map(|n| n.starts_with("bank") && n.ends_with(".cfg"))
                .unwrap_or(false)
        })
        .collect();
    cfg_paths.sort();

    let mut parsed: Vec<BankCfg> = Vec::new();
    for p in &cfg_paths {
        match load_bank_cfg(p) {
            Ok(mut c) => {
                // v2_regen derives the bank from the `bankXX.cfg` filename, not
                // the `bank =` directive. Mirror that authority so the passes
                // (which read cfg.bank) key off the same bank Python uses.
                if let Some(b) = filename_bank(p) {
                    c.bank = b as i32;
                }
                parsed.push(c);
            }
            Err(e) => eprintln!("  parse-fail {}: {e}", p.display()),
        }
    }

    // Build name_map + cross_bank_names (mirrors v2_regen first pass).
    let mut name_map: HashMap<u32, String> = HashMap::new();
    let mut cross_bank_names: HashMap<u32, Vec<NameDecl>> = HashMap::new();
    for cfg in &parsed {
        let bank = cfg.bank as u32;
        for entry in &cfg.entries {
            if let Some(name) = &entry.name {
                name_map.insert((bank << 16) | (entry.start & 0xFFFF), name.clone());
            }
        }
        for nd in &cfg.names {
            let addr = nd.addr_24 & 0xFFFFFF;
            name_map.insert(addr, nd.name.clone());
            cross_bank_names.entry((addr >> 16) & 0xFF).or_default().push(nd.clone());
        }
    }

    // Process-global decode state: inline-skip map (`if n:`) + aggregate reloc.
    let mut inline_skip: HashMap<u32, i32> = HashMap::new();
    for cfg in &parsed {
        let bank = cfg.bank as u32;
        for entry in &cfg.entries {
            if let Some(n) = entry.inline_skip {
                if n != 0 {
                    inline_skip.insert((bank << 16) | (entry.start & 0xFFFF), n);
                }
            }
        }
    }
    let mut reloc_regions: Vec<RelocRegion> = Vec::new();
    for cfg in &parsed {
        reloc_regions.extend(cfg.reloc_regions.iter().copied());
    }

    // The env tail_call + dispatch-helper-discovery decode under (reloc +
    // global inline skip only; no dispatch_helpers / callee_exit_mx / data /
    // indirect — matching the Python passes' kwarg-less decode_function calls).
    let basic_env = DecodeEnv {
        reloc_regions: Some(&reloc_regions),
        global_inline_skip: Some(&inline_skip),
        ..Default::default()
    };

    // 1. wrapper
    let wrapper_fixes = wrapper_detect_and_route(&parsed, &mut name_map, &mut cross_bank_names, &rom);
    // 2. tail_call
    let tail_call_fixes = tail_call_detect_and_route(&mut parsed, &rom, &basic_env);
    // 3. pha_rts
    let pha_rts_fixes = pha_rts_detect_and_route(&mut parsed, &rom);
    // 4. dispatch-helper discovery
    let dispatch_helpers = discover_dispatch_helpers(&parsed, &rom, &basic_env);
    // 5. exit_mx
    let exit_mx_fixes =
        exit_mx_detect_and_route(&mut parsed, &rom, &helpers_as_hashmap(&dispatch_helpers), &reloc_regions);

    // Build output JSON.
    let mut banks = Map::new();
    for cfg in &parsed {
        let bank = cfg.bank as u32;
        let mut entries: Vec<(u32, String, Value)> = cfg
            .entries
            .iter()
            .map(|e| (e.start & 0xFFFF, e.name.clone().unwrap_or_default(), entry_obj(e)))
            .collect();
        entries.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1)));
        let entries_v: Vec<Value> = entries.into_iter().map(|(_, _, v)| v).collect();

        let mut pv: Vec<[u32; 6]> = cfg
            .exit_mx_at_per_variant
            .iter()
            .map(|&(b, a, em, ex, m, x)| {
                [b as u32, a & 0xFFFF, em as u32, ex as u32, m as u32, x as u32]
            })
            .collect();
        pv.sort();
        let pv_v: Vec<Value> = pv.iter().map(|t| json!(t)).collect();

        let mut ind: Vec<(u32, u32, Value)> = cfg
            .indirect_dispatch
            .iter()
            .map(|d| (d.site_pc16 & 0xFFFF, d.count, indirect_obj(d)))
            .collect();
        ind.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1)));
        let ind_v: Vec<Value> = ind.into_iter().map(|(_, _, v)| v).collect();

        banks.insert(
            bank.to_string(),
            json!({
                "entries": entries_v,
                "exit_mx_at_per_variant": pv_v,
                "indirect_dispatch": ind_v,
            }),
        );
    }

    let mut helpers_v = Map::new();
    for (k, v) in &dispatch_helpers {
        helpers_v.insert(k.to_string(), Value::String(v.clone()));
    }

    let out = json!({
        "fix_counts": {
            "wrapper": wrapper_fixes.len(),
            "tail_call": tail_call_fixes.len(),
            "pha_rts": pha_rts_fixes.len(),
            "exit_mx": exit_mx_fixes.len(),
        },
        "dispatch_helpers": Value::Object(helpers_v),
        "banks": Value::Object(banks),
    });

    let s = serde_json::to_string(&out).expect("serialize");
    std::fs::write(&out_path, s).expect("write out");
    eprintln!(
        "wrapper={} tail_call={} pha_rts={} exit_mx={} helpers={} banks={} -> {}",
        wrapper_fixes.len(),
        tail_call_fixes.len(),
        pha_rts_fixes.len(),
        exit_mx_fixes.len(),
        dispatch_helpers.len(),
        parsed.len(),
        out_path,
    );
}

/// exit_mx wants a plain `HashMap<u32,String>`; discovery returns a `BTreeMap`.
fn helpers_as_hashmap(m: &std::collections::BTreeMap<u32, String>) -> HashMap<u32, String> {
    m.iter().map(|(k, v)| (*k, v.clone())).collect()
}
